<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="" xml:lang="">
  <head>
    <title>Generating dynamic reports using R Markdown in RStudio</title>
    <meta charset="utf-8" />
    <meta name="author" content="Momeneh (Sepideh) Foroutan   @S_Foroutan" />
    <meta name="date" content="2019-07-05" />
    <link href="libs/remark-css/default.css" rel="stylesheet" />
    <link href="libs/remark-css/rladies.css" rel="stylesheet" />
    <link href="libs/remark-css/rladies-fonts.css" rel="stylesheet" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Generating dynamic reports using R Markdown in RStudio
### Momeneh (Sepideh) Foroutan <br> <span class="citation">@S_Foroutan</span> <br>
### 2019-07-05

---


---


# About me and R

- Background in Molecular Genetics (BSc and MSc)

- PhD in Computational Cancer Biology

- Self teaching R through on-line courses and forums

--

- Co-founder of [R-Ladies Melbourne](https://www.meetup.com/R-Ladies-Melbourne/events/past/) in Sep 2016 

- President of R-Ladies Melbourne Inc.

&lt;div class="figure"&gt;
&lt;img src="./images/RLadies_meetup.png" alt="https://www.meetup.com/en-AU/rladies-melbourne/" width="70%" /&gt;
&lt;p class="caption"&gt;https://www.meetup.com/en-AU/rladies-melbourne/&lt;/p&gt;
&lt;/div&gt;

---
# R-Ladies 
&lt;div class="figure"&gt;
&lt;img src="./images/RLadies_map.png" alt="https://blog.revolutionanalytics.com/2018/12/women-and-r.html" width="100%" /&gt;
&lt;p class="caption"&gt;https://blog.revolutionanalytics.com/2018/12/women-and-r.html&lt;/p&gt;
&lt;/div&gt;

---
class: inverse, center, middle
# Motivation 

---
# Imagine a project ...
**You are given:**
- Gene expression data for some cancer samples

- A gene experssion signature 

**You are asked to:**
- Find out which samples are more concordant with that signature?

- Communicate the analyses with your colleagues

---
# How I used to do these without R Markdown

I would have:

- a folder with several analysis **scripts**

- a folder with several **figures**

- a folder with several **tables** (e.g. .csv, txt, tsv, etc)

- a **notebook** storing all the notes (rationale of the analyses, methodology, interpretation and descriptions)

- a folder of **papers** that have relevant figures and information. 

--

- **It was always pretty hectic to *communicate* all these information with colleagues, *reproduce* all the results, and *share* my analyses.**

---
# R Markdown was a game changer!

R Markdown is an authoring framework provided by RStudio, which can keep all steps of the analyses together:

- Codes (save and execute)

- Figures and tables

- Methodology, interprations, and descriptions of the analyses

- Link to papers, and images from papers

--

- **It is now much eaier to *communicate* all these information with colleagues, *reproduce* all the results, and *share* my analyses.**
- **Several output formats, and possibilities for static and dynamic (interactive) reports.** 

---
# RStudio
RStudio is an integrated development environment (IDE) for R which includes a console, syntax-highlighting editor, and tools for plotting, history, debugging, etc. Rstudio help you to interact with R more readily.

&lt;img src="./images/RStudio.png" width="100%" style="display: block; margin: auto;" /&gt;
---
# Things you can do using RStudio
- Write, save and run codes

- Generate interactive web application

- Generate high quality reports and documents 

- Making presentation slides

- Version control (Git/Github)

- Many more

--

- [**RStudio website**](https://www.rstudio.com)

- [**Introduction to R and RStudio**](http://ncss-tech.github.io/stats_for_soil_survey/chapters/1_introduction/1_introduction.html)

- [**RStudio cheatsheet**](https://www.rstudio.com/wp-content/uploads/2016/01/rstudio-IDE-cheatsheet.pdf)

---
# Online courses

[**Data Science Specialisation by JHU on Coursera**](https://www.coursera.org/specializations/jhu-data-science)


---
# R Markdown and knitr

R Markdown (.Rmd) combines **R codes** (.R) and **documentatiuon language** (.md) using **knitr**.

**knitr** ...
- inspired by Sweave
- is an R package which works as an engine for dynamic report generation in R. 
- helps us to integrate R codes into other documents (e.g. Markdown, LaTeX, HTML, etc).
- can generate HTML, PDF or Word documents.
- supports other languages, such as Python, Perl, C++, Shell scripts, etc.
- enables reproducible research

---
# R Markdown main sections
YAML header, Markdown text, and code chunks (with outputs).

&lt;img src="./images/RMarkdown_sections.png" width="90%" style="display: block; margin: auto;" /&gt;

---
# Open a new R Markdown
&lt;img src="./images/RMarkdown_start.png" width="100%" style="display: block; margin: auto;" /&gt;

---
# Open a new R Markdown
&lt;img src="./images/RMarkdown_start2.png" width="100%" style="display: block; margin: auto;" /&gt;

---
# Open a new R Markdown
&lt;img src="./images/RMarkdown_start3.png" width="100%" style="display: block; margin: auto;" /&gt;

---
# Open a new R Markdown
&lt;img src="./images/RMarkdown_knit.png" width="100%" style="display: block; margin: auto;" /&gt;

---
# Open a new R Markdown
&lt;img src="./images/RMarkdown_knit2.png" width="100%" style="display: block; margin: auto;" /&gt;

---
# R Notebook

An [**R Notebook**](https://bookdown.org/yihui/rmarkdown/notebook.html#fig:notebook-demo) is an R Markdown document with chunks that can be executed independently and interactively, with output visible immediately beneath the input.
- **R Markdown** (knitr botton): executes and evaluates all code in one go. It can be very time-consuming when we have heavy processing
- **R Notebook** (preview botton): has caching behaviour; it evaluates a code chunk and save it. 

"The immediacy of notebook mode makes it a good choice while authoring the R Markdown document and iterating on code. When you are ready to publish the document, you can share the notebook directly, or render it to a publication format with the Knit button."

---
# R Markdown and R Notebook
.pull-left[
- [R Markdown](https://rmarkdown.rstudio.com)
  + **knit** botton
  + output: **html_document**

&lt;img src="./images/RMarkdown_example.png" width="100%" style="display: block; margin: auto;" /&gt;
]

.pull-right[
- [R Notebook](https://bookdown.org/yihui/rmarkdown/notebook.html) 
  + **preview** botton
  + output: **html_notebook**
  
&lt;img src="./images/Rnotebook_example.png" width="100%" style="display: block; margin: auto;" /&gt;
]

---
# How does it look when we knit?
&lt;div class="figure" style="text-align: center"&gt;
&lt;img src="./images/Rendered_RMarkdown.png" alt="https://github.com/DavisLaboratory/NK_scoring/" width="90%" /&gt;
&lt;p class="caption"&gt;https://github.com/DavisLaboratory/NK_scoring/&lt;/p&gt;
&lt;/div&gt;

---
class: inverse
# Creating the document
.pull-top[
- Modify YAML header
- Structure and format the text
- Insert and modify code chuncks
- Generate interactive tables
- Generate interactive figures
]
.pull-bottom[

&lt;img src="./images/RMarkdown_example2.png" width="100%" style="display: block; margin: auto auto auto 0;" /&gt;

]
---
# YAML header

"YAML (YAML Ain't Markup Language) is a human-friendly data serialization standard for all programming languages."

YAML header allows us to modify the output of the domunent; YAML section can evaluate R expressions.
By modifying the YAML header, we can add/change:
- Output format
- Table of content (toc)
- Tabbed sections 
- Global figure options (width and height)
- Custom CSS
- a lot more

---
# YAML header
.pull-top[
&lt;img src="./images/YAML.png" width="70%" style="display: block; margin: auto;" /&gt;
]
.pull-bottom[
&lt;img src="./images/YAML_code_hide.png" width="70%" style="display: block; margin: auto;" /&gt;
]

---
# YAML header
.pull-top[
&lt;img src="./images/YAML.png" width="70%" style="display: block; margin: auto;" /&gt;
]
.pull-bottom[
&lt;img src="./images/YAML_code_show.png" width="70%" style="display: block; margin: auto;" /&gt;
]

---
# Structure and modify the text

.pull-left[
- Headings and sub-headings are generated using **#**:

`# Heading 1`
# Heading 1

`## Heading 2` 
## Heading 2
]

.pull-right[
- Italic texts are generated using * *:

`*This is italic!*` 

*This is italic!*

- Bold texts are generated using ** **:

`**This is bold!**` 

**This is bold!**
]

---
# Structure and modify the text

- Add hyperlinks using `[your_text](your_url)`; for example:

`[R Markdown documnetation](https://rmarkdown.rstudio.com)` will make [R Markdown documnetation](https://rmarkdown.rstudio.com) clickable, which opens up the corresponding webpage for the documentation.

- Use **single backticks** as wrappers to change the font to make the code, package names, etc different from other plain texts.

- Add bullet points using **minus** and **plus** signs as well as tab.

.pull-left[
For example:

`-` First point

Tab `+` class A

Tab `+` class B

`-` Second point

]

.pull-right[
Will result in:

- First point
  + class A
  + class B
- Second point

]





---
# Code chunks

&lt;img src="./images/code_chunks1.png" width="65%" style="display: block; margin: auto;" /&gt;

--

&lt;img src="./images/code_chunks2.png" width="65%" style="display: block; margin: auto;" /&gt;

--

&lt;img src="./images/code_chunks3.png" width="65%" style="display: block; margin: auto;" /&gt;

---
# Insert/run code chunks

.pull-left[
&lt;img src="./images/Rnotebook_example.png" width="100%" style="display: block; margin: auto;" /&gt;
]

--

.pull-right[
&lt;img src="./images/code_chunks_bottons1.png" width="80%" style="display: block; margin: auto;" /&gt;


- In Mac: Command + Option + i
- In Windows: Ctrl + Alt + i

]

---
# Insert/run code chunks
.pull-left[
&lt;img src="./images/Rnotebook_example.png" width="100%" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;img src="./images/code_chunks_bottons2.png" width="80%" style="display: block; margin: auto;" /&gt;
]

---
# Interactive tables using DT package

.pull-left[
- Interface to the DataTable javascript library
- Very easy-to-use: `datatable(df)`
- filtering, paging, sorting, formatting the tables, etc.
- [DT documentation](https://rstudio.github.io/DT/)
]

--

.pull-right[
&lt;img src="./images/DT_example_code.png" width="100%" style="display: block; margin: auto auto auto 0;" /&gt;
]

&lt;img src="./images/DT_example_plot.png" width="80%" style="display: block; margin: auto;" /&gt;

---
# Interactive plots using plotly package

- Plotly can generate 2D and 3D plots, as well as animations

- It is possible to zoom, pan, label, and toggle between items in the legend 

- Save static image functionality

- Configurable tooltips

- Very easy-to-use with **ggplot**: `ggplotly(ggplot_object)`

- [Plotly documentation](https://plot.ly/r/)

---
class: inverse
# Back to our theoritical project...

Use samples' gene expression data and a gene set scoring method to score samples against gene sets and identify those that are more concordant with a given signature. 

--

&lt;br&gt;

.pull-left[
&lt;img src="./images/Singscore_logo.png" width="70%" style="display: block; margin: auto;" /&gt;
]

.pull-right[
&lt;br&gt;
- Rank-based method

- Single-sample approach 

- R/Bioconductor package

- Interactive plots
]


---
# The singscore method
&lt;img src="./images/singscore_workflow.png" width="91%" style="display: block; margin: auto;" /&gt;

&lt;br&gt;
&lt;small&gt;[Single sample scoring of molecular phenotypes. Foroutan M, Bhuva D, et al. *BMC Bioinformatics*](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/s12859-018-2435-4)&lt;/small&gt;

---
# Visualisations by `singscore`


&lt;img src="./images/RankDensity.png" width="70%" style="display: block; margin: auto;" /&gt;
--
&lt;img src="./images/ScoreDispersion.png" width="70%" style="display: block; margin: auto;" /&gt;
&lt;small&gt;[Using singscore to predict mutations in AML from transcriptomic signatures. Bhuva D. et al. *f1000 research*](https://bioconductor.org/packages/release//workflows/vignettes/SingscoreAMLMutations/inst/doc/workflow_transcriptional_mut_sig.html).&lt;/small&gt;

---
# Visualisations by `singscore`

&lt;img src="./images/ScoreLandscape.png" width="50%" style="display: block; margin: auto;" /&gt;
&lt;small&gt;[Using singscore to predict mutations in AML from transcriptomic signatures. Bhuva D. et al. *f1000 research*](https://bioconductor.org/packages/release//workflows/vignettes/SingscoreAMLMutations/inst/doc/workflow_transcriptional_mut_sig.html).&lt;/small&gt;

---
class: inverse, middle, center
# Let's look at our R Markdown report 

[**Generating dynamic reports in R Markdown - an example using singscore**](file:///Users/foroutanm/Documents/presentations/conferences/WinterSchool_2019_Brisbane/Slides/analysis/winterSchool_singscore.html)

---
# Resources I used for this presentation

- [**RStudio website**](https://www.rstudio.com)
- [**R Markdown documentation**](https://rmarkdown.rstudio.com)
- [**R Notebook documentation**](https://bookdown.org/yihui/rmarkdown/notebook.html#fig:notebook-demo)
- [**Baby one more time - Reproducibility in R and when to pull in the big guns** by *Lavinia Gordon*](https://mrslaviniag.github.io/xaringan_presentations/rladies_presentation_22_May_2019.html)

- [**RLadies presentation Ninja** by *Alison Presmanes Hill*](https://alison.rbind.io/slides/rladies-demo-slides.html)
- [**Making slides in R Markdown** by *Alison Hill*](https://arm.rbind.io/slides/xaringan.html)

---
class: inverse, center, middle
# Thank you!
    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"highlightStyle": "github",
"highlightLines": true,
"countIncrementalSlides": false
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function(time) {
  var d2 = function(number) {
    return ('0' + number).slice(-2); // left-pad 0 to minutes/seconds
  },

  time_format = function(total) {
    var secs = Math.abs(total) / 1000;
    var h = Math.floor(secs / 3600);
    var m = Math.floor(secs % 3600 / 60);
    var s = Math.round(secs % 60);
    var res = d2(m) + ':' + d2(s);
    if (h > 0) res = h + ':' + res;
    return res;  // [hh:]mm:ss
  },

  slide_number_div = function(i) {
    return document.getElementsByClassName('remark-slide-number').item(i);
  },

  current_page_number = function(i) {
    return slide_number_div(i).firstChild.textContent;  // text "i / N"
  };

  var timer = document.createElement('span'); timer.id = 'slide-time-left';
  var time_left = time, k = slideshow.getCurrentSlideIndex(),
      last_page_number = current_page_number(k);

  setInterval(function() {
    time_left = time_left - 1000;
    timer.innerHTML = ' ' + time_format(time_left);
    if (time_left < 0) timer.style.color = 'red';
  }, 1000);

  slide_number_div(k).appendChild(timer);

  slideshow.on('showSlide', function(slide) {
    var i = slide.getSlideIndex(), n = current_page_number(i);
    // reset timer when a new slide is shown and the page number is changed
    if (last_page_number !== n) {
      time_left = time; last_page_number = n;
      timer.innerHTML = ' ' + time_format(time); timer.style.color = null;
    }
    slide_number_div(i).appendChild(timer);
  });
})(50000);</script>

<script>
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
